**This is my code result for Data Mining**

In [1]:
import numpy as np
import pandas as pd

In [2]:
Jan=pd.read_csv('JANUARY.csv')
Feb=pd.read_csv('FEBRUARY.csv')
MARCH=pd.read_csv('MARCH.csv')
APRIL=pd.read_csv('APRIL.csv')
MAY=pd.read_csv('MAY.csv')
JUNE=pd.read_csv('JUNE.csv')

C:\ProgramData\Anaconda3\lib\site-packages\IPython\core\interactiveshell.py:3057: DtypeWarning: Columns (77,84,85) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)
C:\ProgramData\Anaconda3\lib\site-packages\IPython\core\interactiveshell.py:3057: DtypeWarning: Columns (48,77,84) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [3]:
REPORT_DATA=pd.concat([Jan,Feb,MARCH,APRIL,MAY,JUNE])

C:\Users\ogunniran siji\AppData\Roaming\Python\Python37\site-packages\ipykernel_launcher.py:1: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  """Entry point for launching an IPython kernel.


In [6]:
REPORT_DATA.to_csv('REPORT_DATA')

**Question 1**

The airline that had most number of travels between January - March 2019

In [27]:
DATA = pd.concat([Jan,Feb,MARCH])

In [29]:
i=DATA['OP_CARRIER_AIRLINE_ID'].unique()
i

array([20363, 19393, 20368, 20397, 20409, 20416, 20304, 19805, 20398,
       19977, 19790, 20378, 20452, 19930, 20436, 19690, 20366],
      dtype=int64)

In [30]:
for x in i:
    data=DATA['FLIGHTS'][DATA['OP_CARRIER_AIRLINE_ID']==x].sum()
    print(data)
    

60880.0
330225.0
24294.0
69080.0
72768.0
46030.0
194934.0
228103.0
75751.0
144288.0
225381.0
53701.0
76818.0
61466.0
28242.0
19689.0
37584.0


From the results 19393 has the highest number of flights of 330225. The Airline is **Southwest Airlines Co.: WN**

**Question 2**

What is the total number of Airline Travels in the first half of 2019 

In [8]:
REPORT_DATA['FLIGHTS'].sum()

3634338.0

**Question 3**

What is the Total AirTime by all Airlines in May 2019 

In [9]:
MAY['AIR_TIME'].sum()

69297104.0

For Minute to hour Conversion 69297104 minutes = 1154951.73333 hours

**Question 4**

What city do you consider as the busiest city in April 2019 

In [10]:
APRIL['ORIGIN_CITY_NAME'].value_counts()[:5]

Chicago, IL              34396
Atlanta, GA              32940
New York, NY             25094
Dallas/Fort Worth, TX    23933
Houston, TX              19936
Name: ORIGIN_CITY_NAME, dtype: int64

In [11]:
APRIL['DEST_CITY_NAME'].value_counts()[:5]

Chicago, IL              34393
Atlanta, GA              32941
New York, NY             25095
Dallas/Fort Worth, TX    23930
Houston, TX              19941
Name: DEST_CITY_NAME, dtype: int64

From the Origin and departure results Chicago is the most busiest city

**Question 5**

What airline covered the most distance in June 2019 

In [12]:
i=JUNE['OP_CARRIER_AIRLINE_ID'].value_counts()
i=i.index
i

Int64Index([19393, 19790, 19805, 20304, 19977, 20398, 20452, 20397, 20409,
            19930, 20363, 20378, 20416, 20436, 20368, 20366, 19690],
           dtype='int64')

In [13]:
for x in i:
    june=JUNE['DISTANCE'][JUNE['OP_CARRIER_AIRLINE_ID']==x].sum()
    print(june)
    

85246573.0
79641435.0
80947454.0
35631686.0
65963246.0
12910059.0
16804684.0
9877331.0
26972049.0
30800766.0
10009285.0
11066389.0
17576612.0
11879495.0
9283200.0
4389648.0
5316126.0


From the results The Airline Code **19393** has a total Distance of **85246573.0** which is the highest distance. 
    The Airline of the code is **Southwest Airlines Co.: WN**

**Question 6**

What is the Tail Number of the aircraft that covered the most distance in Q2 of 2019 

In [4]:
#Get the Unique values in the second quarter
TAIL=(REPORT_DATA['TAIL_NUM'][REPORT_DATA['QUARTER']==2]).unique()
TAIL_NUM=TAIL

In [7]:
#Converts it to a dataframe
TAIL_NUMS=pd.DataFrame(data=TAIL_NUM)

In [8]:
TAIL_NUMS.columns=['TAIL']

In [10]:
TAIL_NUMS.tail()

TAIL
5554  N116DU
5555  N27958
5556  N220HA
5557  N2341U
5558  N272AY

In [11]:
#Second Quater Consists of April, May, June So we join them together
Second_Quater=pd.concat([APRIL,MAY,JUNE])

C:\Users\ogunniran siji\AppData\Roaming\Python\Python37\site-packages\ipykernel_launcher.py:2: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  


In [12]:
#Iterate through each value to discover the total distance sum of each
for i in TAIL_NUM:

    TOTAL_DIST=Second_Quater['DISTANCE'][Second_Quater['TAIL_NUM']==i].sum()
        
    print(TOTAL_DIST)

437437.0
416662.0
416145.0
449392.0
422161.0
405298.0
379006.0
423591.0
364063.0
420142.0
409808.0
379505.0
392934.0
440454.0
450530.0
464178.0
443055.0
438822.0
396970.0
388269.0
422128.0
430302.0
397149.0
443883.0
411758.0
449585.0
414246.0
430520.0
430613.0
393872.0
402866.0
395940.0
381406.0
473842.0
409873.0
410368.0
407521.0
443304.0
447371.0
415065.0
414614.0
425979.0
400787.0
451696.0
433975.0
333391.0
373446.0
477155.0
457165.0
431570.0
377765.0
437727.0
437137.0
427711.0
451793.0
455038.0
470080.0
423119.0
416501.0
349872.0
413403.0
362226.0
369545.0
413740.0
383870.0
384286.0
400012.0
398692.0
384108.0
339968.0
352880.0
359080.0
288437.0
296709.0
312233.0
305186.0
355770.0
327563.0
303133.0
308348.0
331130.0
295097.0
309695.0
315257.0
355499.0
331689.0
457674.0
395656.0
414246.0
400191.0
453331.0
455407.0
429066.0
426563.0
436898.0
397269.0
387634.0
386527.0
406541.0
416497.0
328597.0
378901.0
435715.0
438901.0
330399.0
368808.0
379972.0
349063.0
312863.0
204584.0
338491.0
3

205560.0
304222.0
405067.0
274471.0
313868.0
304214.0
291309.0
293022.0
181705.0
310280.0
370659.0
225069.0
331769.0
363635.0
267109.0
236201.0
303245.0
280677.0
282275.0
446348.0
555527.0
228210.0
308574.0
262592.0
306377.0
548339.0
539162.0
542650.0
273765.0
328518.0
281077.0
305554.0
311268.0
292067.0
323935.0
551921.0
320167.0
372165.0
279180.0
353241.0
144453.0
219905.0
592006.0
475939.0
311382.0
295259.0
558408.0
575250.0
353759.0
329874.0
320756.0
281636.0
285522.0
374169.0
303220.0
261634.0
343336.0
575520.0
598286.0
316244.0
571284.0
364331.0
357317.0
384037.0
257931.0
554580.0
553653.0
247905.0
355614.0
560119.0
581559.0
284074.0
572129.0
310198.0
242116.0
229186.0
323410.0
186544.0
316116.0
295181.0
338689.0
298749.0
248988.0
249699.0
331120.0
524216.0
165745.0
336443.0
270270.0
333873.0
345237.0
278339.0
320609.0
320540.0
396361.0
309847.0
572512.0
284981.0
583407.0
324726.0
240973.0
353053.0
298698.0
352997.0
334965.0
270278.0
273101.0
534011.0
502711.0
381214.0
298002.0
3

341669.0
266230.0
230966.0
249961.0
348117.0
316215.0
374682.0
384041.0
408106.0
399417.0
290993.0
393366.0
380986.0
358699.0
423755.0
440511.0
400012.0
108114.0
431582.0
134543.0
122984.0
72950.0
114348.0
213047.0
118245.0
311526.0
293221.0
256047.0
241597.0
261983.0
258232.0
242681.0
208766.0
332656.0
264787.0
305578.0
143591.0
273385.0
299190.0
286571.0
252484.0
184117.0
206402.0
346637.0
302815.0
344525.0
347709.0
308873.0
236206.0
300514.0
340692.0
297963.0
284418.0
118211.0
130703.0
133690.0
131669.0
103398.0
95552.0
119251.0
146207.0
95477.0
126623.0
129217.0
88912.0
271417.0
105129.0
113561.0
211569.0
291933.0
19115.0
202750.0
181423.0
318006.0
229092.0
277054.0
348344.0
266807.0
311639.0
259386.0
198464.0
173799.0
196445.0
297688.0
222782.0
290373.0
236429.0
284477.0
215615.0
287836.0
314663.0
322906.0
333266.0
317552.0
313465.0
259986.0
212759.0
270261.0
276830.0
329109.0
176679.0
158388.0
301914.0
187741.0
110283.0
189665.0
299734.0
264153.0
120263.0
171808.0
99817.0
271879.

346352.0
251140.0
296994.0
234090.0
334573.0
270959.0
250059.0
374513.0
269524.0
148155.0
294095.0
205216.0
306302.0
343554.0
503445.0
281801.0
162463.0
221466.0
328410.0
215846.0
231375.0
220984.0
269335.0
297287.0
43128.0
337251.0
381256.0
278191.0
224210.0
347089.0
231438.0
159679.0
232173.0
62087.0
210111.0
398115.0
260545.0
212117.0
231748.0
510267.0
386600.0
484055.0
216567.0
252683.0
259310.0
288318.0
271359.0
137363.0
331844.0
309889.0
221295.0
413502.0
377006.0
518973.0
299782.0
397681.0
357321.0
337700.0
355003.0
358356.0
250634.0
474576.0
324351.0
320010.0
373919.0
291923.0
304319.0
212613.0
344832.0
319283.0
452372.0
394869.0
387813.0
302953.0
427941.0
321591.0
248863.0
198088.0
350484.0
432874.0
220790.0
203022.0
403069.0
226038.0
391659.0
328246.0
222745.0
456120.0
211354.0
214729.0
315017.0
234961.0
246477.0
355681.0
266145.0
195473.0
230087.0
294170.0
325371.0
479183.0
329654.0
403436.0
161614.0
172168.0
335485.0
44550.0
357362.0
230951.0
317555.0
33695.0
295257.0
30983

13531.0
27558.0
130516.0
337889.0
451046.0
281912.0
37115.0
59514.0
261024.0
17600.0
327583.0
219775.0
19543.0
263043.0
66289.0
111248.0
343299.0
21986.0
6192.0
23821.0
302539.0
12966.0
365600.0
38258.0
26880.0
23188.0
15169.0
14275.0
98108.0
20103.0
257956.0
153279.0
265079.0
257073.0
28732.0
19835.0
42671.0
60484.0
222678.0
70642.0
10949.0
5676.0
5516.0
194269.0
154210.0
318715.0
20195.0
34576.0
12176.0
74998.0
25679.0
342024.0
248203.0
324376.0
257186.0
307429.0
251546.0
249388.0
291656.0
290148.0
297591.0
279519.0
252805.0
260690.0
305600.0
321891.0
363470.0
237154.0
318949.0
288116.0
386148.0
300341.0
284007.0
255033.0
304708.0
337353.0
276945.0
355782.0
302238.0
323879.0
272308.0
229125.0
333110.0
351845.0
294684.0
354424.0
266025.0
388202.0
269894.0
284139.0
232634.0
258686.0
322462.0
306512.0
271372.0
392887.0
265606.0
349595.0
326561.0
314525.0
341899.0
262953.0
326575.0
277058.0
25198.0
353838.0
318802.0
299223.0
331692.0
313359.0
362421.0
308256.0
272946.0
347907.0
326053.0


171918.0
189552.0
172948.0
181353.0
184256.0
255477.0
239085.0
239468.0
246611.0
239150.0
212945.0
238871.0
212982.0
251065.0
232527.0
236237.0
213695.0
228811.0
229132.0
183625.0
205347.0
237300.0
220441.0
225431.0
203781.0
190739.0
185631.0
181899.0
192571.0
189525.0
205802.0
171330.0
172861.0
271884.0
248077.0
227194.0
231899.0
151798.0
232918.0
181051.0
205218.0
218928.0
208536.0
197179.0
230408.0
145482.0
221938.0
225136.0
253624.0
212354.0
185377.0
191219.0
238848.0
222720.0
189944.0
179501.0
227806.0
197154.0
207071.0
209806.0
185405.0
233348.0
238313.0
192052.0
179559.0
126996.0
205961.0
180155.0
167384.0
205445.0
116167.0
195023.0
199532.0
314382.0
288650.0
246675.0
209609.0
200623.0
239517.0
272960.0
262803.0
186884.0
259705.0
254846.0
240099.0
272405.0
270710.0
265712.0
263661.0
251855.0
264452.0
262812.0
257561.0
257429.0
214302.0
268383.0
255304.0
241378.0
259104.0
167121.0
239618.0
186158.0
238139.0
246543.0
267244.0
269611.0
44068.0
257591.0
257398.0
271717.0
183542.0
25

97020.0
33913.0
69786.0
68265.0
44809.0
35344.0
30327.0
119307.0
25042.0
32965.0
88101.0
2758.0
14946.0
81655.0
54549.0
3227.0
9368.0
4684.0
3894.0
47998.0
17273.0
13586.0
1358.0
8160.0
2758.0
100908.0
4838.0
3152.0


In [16]:
#The results are saved as a csv which I read again
sumx=pd.read_csv('SUMX.csv')

In [17]:
sumx.tail()

SUM
5554    8160
5555    2758
5556  100908
5557    4838
5558    3152

In [18]:
#Join the Unique values and Sum together
name=pd.concat([TAIL_NUMS,sumx],axis=1)

In [19]:
name.head()

TAIL     SUM
0  N646NK  437437
1  N615NK  416662
2  N633NK  416145
3  N655NK  449392
4  N605NK  422161

In [21]:
#Gets The maximum Sum and discover the tail number
[name['SUM'].max()]

[598286]

In [22]:
name[name['SUM']==598286]

TAIL     SUM
969  N934JB  598286

From the Results the The Tail Number N934JB has the distance sum of 598286 which is the highest

**Question 7**

Which state recorded the most delays in the 2nd Quarter of 2019

In [51]:
#Calculate Total delay as sum of Arrival+Departure Delay
Second_Quater['TOT_DELAY']=Second_Quater['ARR_DELAY']+Second_Quater['DEP_DELAY']

In [52]:
#Gets the Unique Origin State
Sec_Quat=Second_Quater['ORIGIN_STATE_NM'].unique()
Sec_Qua=pd.DataFrame(Sec_Quat)

In [53]:
Sec_Qua.columns=['ORIGIN_STATE']

In [54]:
Sec_Qua.head()

ORIGIN_STATE
0     New York
1     Maryland
2     Michigan
3      Florida
4       Nevada

In [55]:
#Calculate the Total delay for each Unique Origin State
for i in Sec_Quat:
     
    j=(Second_Quater['TOT_DELAY'][Second_Quater['ORIGIN_STATE_NM']==i]).sum()
        
    print(j)

2277710.0
467124.0
791989.0
3717306.0
844257.0
2733574.0
1308363.0
354162.0
863145.0
1893966.0
526775.0
5857487.0
3159940.0
171299.0
81242.0
2129442.0
387890.0
1455057.0
890304.0
273823.0
509131.0
1878340.0
908649.0
761049.0
94464.0
78762.0
478120.0
208704.0
253995.0
111734.0
215708.0
765934.0
33120.0
125004.0
97961.0
113440.0
331262.0
191223.0
21060.0
219118.0
106046.0
103738.0
49815.0
1099.0
43443.0
115754.0
13294.0
73826.0
59230.0
-61692.0
50523.0
-3341.0


In [56]:
#The results are saved as a csv which I read again
TOTAL=pd.read_csv('TOT_DELAY.csv')
TOTAL.head()

TOT_STATE_DEL
0        2277710
1         467124
2         791989
3        3717306
4         844257

In [58]:
FIN=pd.concat([Sec_Qua,TOTAL],axis=1)

In [59]:
FIN['TOT_STATE_DEL'].max()

5857487

In [60]:
#Locates the State with the maximum Total delay
FIN.loc[FIN['TOT_STATE_DEL']==5857487]

ORIGIN_STATE  TOT_STATE_DEL
11        Texas        5857487

From the results Texas has the highest total delays of 5857487

**Question 8**

What is the daily average Taxi-out time in New York in February 2019 

In [61]:
Feb.head()

YEAR  QUARTER  MONTH  DAY_OF_MONTH  DAY_OF_WEEK     FL_DATE  \
0  2019        1      2             4            1  2019-02-04   
1  2019        1      2             5            2  2019-02-05   
2  2019        1      2             6            3  2019-02-06   
3  2019        1      2             7            4  2019-02-07   
4  2019        1      2             8            5  2019-02-08   

  OP_UNIQUE_CARRIER  OP_CARRIER_AIRLINE_ID OP_CARRIER TAIL_NUM  ...  \
0                AA                  19805         AA   N989AN  ...   
1                AA                  19805         AA   N346PR  ...   
2                AA                  19805         AA   N872NN  ...   
3                AA                  19805         AA   N316PF  ...   
4                AA                  19805         AA   N855NN  ...   

   DIV4_TAIL_NUM  DIV5_AIRPORT  DIV5_AIRPORT_ID  DIV5_AIRPORT_SEQ_ID  \
0            NaN           NaN              NaN                  NaN   
1            NaN           NaN              NaN                  NaN   
2            NaN           NaN              NaN                  NaN   
3            NaN           NaN              NaN                  NaN   
4            NaN           NaN              NaN                  NaN   

  DIV5_WHEELS_ON DIV5_TOTAL_GTIME DIV5_LONGEST_GTIME  DIV5_WHEELS_OFF  \
0            NaN              NaN                NaN              NaN   
1            NaN              NaN                NaN              NaN   
2            NaN              NaN                NaN              NaN   
3            NaN              NaN                NaN              NaN   
4            NaN              NaN                NaN              NaN   

  DIV5_TAIL_NUM  Unnamed: 109  
0           NaN           NaN  
1           NaN           NaN  
2           NaN           NaN  
3           NaN           NaN  
4           NaN           NaN  

[5 rows x 110 columns]

In [62]:
Feb2=pd.DataFrame(Feb, columns=['DAY_OF_MONTH','ORIGIN_STATE_NM','TAXI_OUT'])

In [63]:
Feb2.head()

DAY_OF_MONTH ORIGIN_STATE_NM  TAXI_OUT
0             4         Florida      21.0
1             5         Florida      22.0
2             6         Florida      17.0
3             7         Florida      13.0
4             8         Florida      21.0

In [65]:
Feb2['ORIGIN_STATE_NM'].unique()

array(['Florida', 'Massachusetts', 'California', 'Hawaii', 'Arizona',
       'Louisiana', 'North Carolina', 'Georgia', 'Texas', 'Nevada',
       'Pennsylvania', 'Tennessee', 'Illinois', 'New Jersey', 'New York',
       'Minnesota', 'Oklahoma', 'Puerto Rico', 'Missouri', 'Virginia',
       'Maryland', 'South Carolina', 'Arkansas', 'Colorado', 'Michigan',
       'Kansas', 'Nebraska', 'Utah', 'U.S. Virgin Islands', 'Oregon',
       'New Mexico', 'Connecticut', 'Alabama', 'Idaho', 'Ohio',
       'Kentucky', 'Iowa', 'Washington', 'Indiana', 'New Hampshire',
       'Wisconsin', 'Rhode Island', 'Maine', 'Vermont', 'Mississippi',
       'West Virginia', 'Montana', 'North Dakota', 'South Dakota',
       'Wyoming', 'Alaska',
       'U.S. Pacific Trust Territories and Possessions'], dtype=object)

In [66]:
#Gets the data for new york
Feb2=Feb2[Feb2['ORIGIN_STATE_NM']=='New York']

In [67]:
#Finds  the average for each day departure
FEB2=Feb2.groupby('DAY_OF_MONTH').mean().head()

In [68]:
#Finds the average Taxi-Out Time for each
FEB2['TAXI_OUT'].mean()

19.259882290853366

The Daily Average Taxi out Time in New york is 19.25

**Question 9**

How many airlines with New York destinations were diverted in Q1 2019

In [3]:
#Since First Quater is Jan, Feb and March we join them together

DIV=pd.concat([Jan,Feb,MARCH])

In [4]:
DIV2=DIV[DIV['DEST_STATE_NM']=='New York']

In [5]:
DIV2.head()

YEAR  QUARTER  MONTH  DAY_OF_MONTH  DAY_OF_WEEK     FL_DATE  \
56   2019        1      1             1            2  2019-01-01   
57   2019        1      1             2            3  2019-01-02   
154  2019        1      1             2            3  2019-01-02   
155  2019        1      1             3            4  2019-01-03   
156  2019        1      1             4            5  2019-01-04   

    OP_UNIQUE_CARRIER  OP_CARRIER_AIRLINE_ID OP_CARRIER TAIL_NUM  ...  \
56                 9E                  20363         9E   N136EV  ...   
57                 9E                  20363         9E   N341PQ  ...   
154                9E                  20363         9E   N337PQ  ...   
155                9E                  20363         9E   N607LR  ...   
156                9E                  20363         9E   N929XJ  ...   

     DIV4_TAIL_NUM  DIV5_AIRPORT  DIV5_AIRPORT_ID  DIV5_AIRPORT_SEQ_ID  \
56             NaN           NaN              NaN                  NaN   
57             NaN           NaN              NaN                  NaN   
154            NaN           NaN              NaN                  NaN   
155            NaN           NaN              NaN                  NaN   
156            NaN           NaN              NaN                  NaN   

    DIV5_WHEELS_ON DIV5_TOTAL_GTIME DIV5_LONGEST_GTIME  DIV5_WHEELS_OFF  \
56             NaN              NaN                NaN              NaN   
57             NaN              NaN                NaN              NaN   
154            NaN              NaN                NaN              NaN   
155            NaN              NaN                NaN              NaN   
156            NaN              NaN                NaN              NaN   

    DIV5_TAIL_NUM  Unnamed: 109  
56            NaN           NaN  
57            NaN           NaN  
154           NaN           NaN  
155           NaN           NaN  
156           NaN           NaN  

[5 rows x 110 columns]

In [6]:
DIV2['DIVERTED'].sum()

260.0

From the results we see that the Total Diverted Airlines in the first Quarter of 2019 is 260

**Question 10**

Which of these delays is more rampant in Washington in the first half of 2019

In [8]:
#Gets Wahington Data and Calculates total delay for each Category
REPORT_TABLE=REPORT_DATA[REPORT_DATA['ORIGIN_STATE_NM']=='Washington']

In [9]:
REPORT_TABLE['CARRIER_DELAY'].sum()

230277.0

In [10]:
REPORT_TABLE['WEATHER_DELAY'].sum()

59108.0

In [11]:
REPORT_TABLE['NAS_DELAY'].sum()

238556.0

In [12]:
REPORT_TABLE['SECURITY_DELAY'].sum() 

988.0

In [13]:
REPORT_TABLE['LATE_AIRCRAFT_DELAY'].sum() 

242206.0

From the Results LATE_AIRCRAFT_DELAY is the most rampant in Washington with the Total of 242206.0